### PR TITLE
Update to latest FPGA bitstream

### PR DESCRIPTION
**Pull request by desowin**
Use latest FPGA bitstream from upstream commit 4d9263f64c58 ("Do not wait indefinitely for host sdram data burst"). Bitstream changes:
  * Report untruncated packet size to host
  * Set maximum captured packet size to 1027
  * Send timestamp delta instead of actual timestamp
  * Do not send empty flags byte
  * Use packet FIFO to avoid unnecessary producer waits
  * Do not wait indefinitely for host sdram data burst

The protocol changes are not backwards compatible, however considering that application always loads the bitstream before using there is no real reason to stick to the (essentially arbitrary) protocol.

This makes timestamps accurate when the bus is suspended during capture and reduces number of overflows when capturing high-speed traffic.

Some of the changes are not yet merged to upstream repository. The PR is open at https://github.com/OpenVizslaTNG/ov_ftdi/pull/18

### CODE REVIEW

**Comment by matwey** [comment]
Thanks!
I need some time to test this by myself, but I don't think that it is a problem since you say that the firmware changes are not yet merged.

**Comment by matwey** [comment]
I wonder if we could have CI for the bit-stream to ensure that the source code is consistent with what is actually loaded in?

**Comment by desowin** [comment]
I think it should be possible to add read-only register with some kind of "protocol version" to identify if the software can work with the gateware. Then the software would read the register and refuse to work if the value is not recognized. I don't think supporting various bitstream revisions have value that would offset the maintenance burden.

The check however would require actual hardware, so it might not really be suitable for CI. Also, without the check the user will simply notice that things fail pretty badly if there is protocol mismatch so it makes the whole idea questionable.

**Comment by matwey** [comment]
Hi,

Frankly speaking, I see some odd behavior with the new firmware. When the capture stopped and then started again, I see error `Cannot start capture: Wrong checksum`. How do you think what could be wrong here?
I use the following test:

```patch
--- a/tools/sample/main.c
+++ b/tools/sample/main.c
@@ -55,6 +55,10 @@ int main(int argc, char** argv) {
                return 1;
        }

+       for (int i=0; i < 1000; ++i) {
+
+       fprintf(stderr, "i = %d\n", i);
+
        ret = ov_capture_start(ov, &p.packet, sizeof(p), &packet_handler, NULL);
        if (ret < 0) {
                fprintf(stderr, "%s: %s\n", "Cannot start capture", ov_get_error_string(ov));
@@ -74,6 +78,8 @@ int main(int argc, char** argv) {

        ov_capture_stop(ov);

+       }
+
        ov_free(ov);

        return 0;
```

**Comment by matwey** [comment]
> I think it should be possible to add read-only register with some kind of "protocol version" to identify if the software can work with the gateware. Then the software would read the register and refuse to work if the value is not recognized. I don't think supporting various bitstream revisions have value that would offset the maintenance burden.
> 
> The check however would require actual hardware, so it might not really be suitable for CI. Also, without the check the user will simply notice that things fail pretty badly if there is protocol mismatch so it makes the whole idea questionable.

I mean more that we provide prebuilt firmware along the source code, and why end-users ever should trust us? It would be great to build the firmware publicly somewhere in GitHub Actions, however it seems to be impossible due to proprietary software from Xilinx.

**Comment by desowin** [comment]
> Frankly speaking, I see some odd behavior with the new firmware. When the capture stopped and then started again, I see error `Cannot start capture: Wrong checksum`. How do you think what could be wrong here?

I don't really use the stop capture, because I just start new ovextcap instance on each capture which just resets the board. However, the `cha_stop_stream()` order is questionable and I wouldn't really expect it to work the way it is now implemented.

Writing 0 to `CSTREAM_CFG` stops producer (it is the `ena` signal passed to `Producer`) so it should be the first step during capture stop. Then the capture should continue to read packets until the packet with `HF0_LAST` bit (0x20) is received. Only when the stream is drained the `SDRAM_HOST_READ_GO` and `SDRAM_SINK_GO` should be written 0. I am not sure about the order of writes to `SDRAM_HOST_READ_GO` and `SDRAM_SINK_GO` though as I only got familiar with Whacker (Producer and Consumer).

**Comment by desowin** [comment]
> I mean more that we provide prebuilt firmware along the source code, and why end-users ever should trust us? It would be great to build the firmware publicly somewhere in GitHub Actions, however it seems to be impossible due to proprietary software from Xilinx.

The 20 GiB required for Xilinx ISE 14.7 is quite problematic, as it pretty much rules out any free CI. Users certainly can request ISE WebPack license and build bitstream themselves, but the fact that the build is not reproducible makes it hard to tell if the bitstream is really the same. I didn't really investigate in what way the bitstreams are different on each subsequent build.

**Comment by desowin** [comment]
> I am not sure about the order of writes to `SDRAM_HOST_READ_GO` and `SDRAM_SINK_GO` though as I only got familiar with Whacker (Producer and Consumer).

`SDRAM_SINK_GO` activates the transfer from Whacker ringbuf (2048 byte memory synthesized in FPGA) to SDRAM and `SDRAM_HOST_READ_GO` allows the host to read from SDRAM. So when stopping capture, `SDRAM_SINK_GO` should be set to 0 before `SDRAM_HOST_READ_GO`.

There is (and always was) an issue related to SDRAM being 16-bit wide and the rest operating on bytes. Modified packet header can change the probability for the last packet to end on even/odd byte boundary, but the probability inevitably depends on captured data stream.

I cannot think of any solution to the 16-bit wide memory problem other than just adding a filler magic byte, repeated twice at the end of capture, where one of the repeats could be discarded by the host. Doing so however would require changes both to gateware and host software.

**Comment by matwey** [comment]
I think it is great opportunity to finally fix the issues with the stream restart. It is very natural to expect that stream can be restarted multiple times without the device reset.

**Comment by desowin** [comment]
Can you handle the host part (in some branch)? With the reordered stop procedure it should either work just fine or hang on the last byte of `HF0_LAST` packet.

To eliminate the hang, Whacker would need to be changed to output even number of bytes during capture session (add filler if needed). Currently this seems to be somewhat complicated but should be doable after my idea to rework Producer/Consumer communication to facilitate NAK transaction filter on FPGA. The rework would make Producer pass all the metadata (timestamp, length, flags, ringbuf offset to packet start) via FIFO and only the actual captured packet data to be passed via ringbuf. After the rework there could be NAK filter in between Producer and Consumer that would forward/discard packets passed to Consumer. It would be Consumer task to generate the 0xA0 header and compute delta timestamp - so it could just keep 1-bit counter related to how many bytes it did output and add a filler byte after processing HF0_LAST packet if needed.

**Comment by matwey** [comment]
> Can you handle the host part (in some branch)? With the reordered stop procedure it should either work just fine or hang on the last byte of `HF0_LAST` packet.

Sure, I've created #31 to discuss the issue.

> 
> To eliminate the hang, Whacker would need to be changed to output even number of bytes during capture session (add filler if needed). Currently this seems to be somewhat complicated but should be doable after my idea to rework Producer/Consumer communication to facilitate NAK transaction filter on FPGA. The rework would make Producer pass all the metadata (timestamp, length, flags, ringbuf offset to packet start) via FIFO and only the actual captured packet data to be passed via ringbuf. After the rework there could be NAK filter in between Producer and Consumer that would forward/discard packets passed to Consumer. It would be Consumer task to generate the 0xA0 header and compute delta timestamp - so it could just keep 1-bit counter related to how many bytes it did output and add a filler byte after processing HF0_LAST packet if needed.

**Comment by desowin** [comment]
Actually, there is slight window during which `HF0_LAST` packet can get lost. I'll fix it and update this PR.

**Comment by desowin** [comment]
> Actually, there is slight window during which `HF0_LAST` packet can get lost. I'll fix it and update this PR.

Now it is fixed, the `HF0_LAST` should always arrive after `CSTREAM_CFG` transition from 1 to 0 (assuming host is acting sensible and not sending back-to-back CSTREAM_CFG change from 0 to 1 then back to 0, i.e. waits for `HF0_FIRST` and `HF0_LAST` before changing `CSTREAM_CFG`)

**Comment by matwey** [comment]
Thank you. I am eager to test it!

File: include/decoder.h [hunk 1]
```diff
@@ -17,15 +17,15 @@ struct packet_decoder {
...
 		NEED_PACKET_MAGIC,
-		NEED_PACKET_FLAGS_LO,
-		NEED_PACKET_FLAGS_HI,
+		NEED_PACKET_FLAGS,
 		NEED_PACKET_LENGTH_LO,
 		NEED_PACKET_LENGTH_HI,
-		NEED_PACKET_TIMESTAMP_LO,
-		NEED_PACKET_TIMESTAMP_ME,
-		NEED_PACKET_TIMESTAMP_HI,
+		NEED_PACKET_TIMESTAMP_BYTE,
```
Comment (matwey) [inline include/decoder.h]: Can we call it just `NEED_PACKET_TIMESTAMP`?

File: include/openvizsla.h [hunk 2]
```diff
@@ -25,9 +25,10 @@ struct
 #endif
 ov_packet {
 	uint8_t  magic;
-	uint16_t flags;
+	uint8_t flags;
 	uint16_t size;
-	uint32_t timestamp : 24;
+	uint16_t orig_size;
```
Comment (matwey) [inline include/openvizsla.h]: I am little unsure about whole `size`/`orig_size` thing, yet I understand the purpose. The idea behind `struct ov_packet` is to have the same memory layout as in the original OV package but with easy access from C. At the same time I see that `->size` is used only once, so I don't think there is much sense to precalculate the value.

Would something like:
```C
inline uint16_t ov_packet_full_size(struct ov_packet* p) {
    return (p->flags & OV_FLAGS_HF0_TRUNC) ? OV_MAX_PACKET_SIZE : p->size;
}

//...

	    .incl_len = ov_packet_full_size(packet),
	    .orig_len = packet->size,
	    .captured = packet->data,

```
works better?

File: include/openvizsla.h [hunk 3]
```diff
@@ -25,9 +25,10 @@ struct
 #endif
 ov_packet {
 	uint8_t  magic;
-	uint16_t flags;
+	uint8_t flags;
 	uint16_t size;
```
Comment (matwey) [inline include/openvizsla.h:29]: Can we also have 
```C
uint8_t timestamp_len : 3;
uint16_t size : 13;
```
?

Comment (desowin) [reply include/openvizsla.h:29]: Why would you need `timestamp_len` later? Note that bitfields are not really portable so you cannot really take any advantage with bitfields. after the cumulative timestamp is calculated it is always 64-bit and the encoded delta timestamp is of no use.

File: include/openvizsla.h (hunk 2 above)
Comment (desowin) [reply include/openvizsla.h]: The function should be named `ov_packet_captured_size` instead of `ov_packet_full_size`. `full` suggests the opposite of what this really is.

File: include/openvizsla.h (hunk 3 above)
Comment (matwey) [reply include/openvizsla.h:29]: Agree

File: include/openvizsla.h (hunk 2 above)
Comment (matwey) [reply include/openvizsla.h]: Let it be `ov_packet_captured_size`

Comment (desowin) [reply include/openvizsla.h]: Where should this funciton be defined at? And where should the flags be? In `openvizsla.h`?

Comment (matwey) [reply include/openvizsla.h]: I think so.

`struct ov_packet` is a part of the public interface, so the flags and the function should be.

**Comment by desowin** [comment]
Note that ovextcap does not currently work as expected. It seems to have been broken in 56c5d37bcdf9b2e82d2bb91557e9a6317849521d

**Comment by matwey** [comment]
> Note that ovextcap does not currently work as expected. It seems to have been broken in [56c5d37](https://github.com/matwey/libopenvizsla/commit/56c5d37bcdf9b2e82d2bb91557e9a6317849521d)

This need to be fixed, then. I'll take care of it.

**Comment by matwey** [comment]
Surprisingly,  ovextcap doesn't work for me even before [56c5d37](https://github.com/matwey/libopenvizsla/commit/56c5d37bcdf9b2e82d2bb91557e9a6317849521d)

**Comment by matwey** [comment]
@desowin I've found another little issue with this new version of firmware.
Even if `SDRAM_HOST_READ_GO` is set to zero, after some period of time, new `0xd0` is received. I suppose this is the new timeout feature. This is no-go: if the device answered that it seen `SDRAM_HOST_READ_GO=0` command it should not send any `0xd0` subsequently.

**Comment by desowin** [comment]
> @desowin I've found another little issue with this new version of firmware.
> Even if `SDRAM_HOST_READ_GO` is set to zero, after some period of time, new `0xd0` is received. I suppose this is the new timeout feature. This is no-go: if the device answered that it seen `SDRAM_HOST_READ_GO=0` command it should not send any `0xd0` subsequently.

I don't think this is new issue. It always was there and what you see is just flushing the pipes *past* the `SDRAM_HOST_READ_GO`. Earlier the gateware simply held that data stuck until stream was restarted (and for obvious reasons the restart failed because there was a lot of buffered data from previous session that did not make it out).

Setting `SDRAM_HOST_READ_GO` to 0 prevents the capture data from going out of SDRAM. The only real use case for controlling this is if you want to arbitrarily cause SDRAM ring buffer stalls. That is, if all other go signals remain active, but `SDRAM_HOST_READ_GO` is set to 0, then eventually SDRAM will get full with new packet data and you will start seeing overflows.

Note that SDRAM never holds the `0xd0` frames. The `0xd0` is generated later in the processing pipeline. The delay you observe is just due to flushing the data that already was in output fifo (i.e. read back from SDRAM) **before** `SDRAM_HOST_READ_GO` went to 0. Every single capture data byte goes the full pipeline, i.e. ULPI -> async fifo -> producer -> whacker ringbuf -> consumer (whole packet can be dropped by filter before reaching consumer) -> SDRAM writer -> SDRAM reader (`SDRAM_HOST_READ_GO` disables the SDRAM reader) -> output fifo -> FTDI

The whole gateware is essentially state machines with internal buffers. If you really want to cleanly stop the whole machinery, you have to clear `CSTREAM_CFG` least significant bit (so no new data is generated at "producer" stage) **and** wait for the host software to fully read packet with `HF0_LAST` bit set (the last packet output by "producer" stage has this bit set). After you read the `HF0_LAST` packet you can be sure that the whole pipeline is clear and there is no stale data lying around.

**Comment by desowin** [comment]
The reason why sdram reader output fifo contents (the `0xd0` frame is generated based on the fifo) appears after register write confirmation is because the register accesses are essentially asynchronous to the main packet stream and the register writes (and reads) are directly writing to FTDI FT.

There is arbitration between the FTDI writers and there has to be enough metadata in the burst writes for the host to separate the two streams that share single transport interface. The `0xd0` wrapper is solely to encapsulate the packet stream, as otherwise it would be really hard to mix the asynchronous confirmations. 

There is no CPU in gateware and thus sequencing the packets would be really complex (and is completely unnecessary in my opinion as it would provide no real benefit).

**Comment by matwey** [comment]
Hi,

Sorry for the delay, I've been busy these days and it took me some time to figure out what is going on.

You are completely right about finite state machines, but this is not how the device behaves with this PR currently.
No matter why we need setting `SDRAM_HOST_READ_GO` to `0`, only thing we have is the serial byte stream to communicate with the device, so the only valid assumption is the following:

```
    TIME ->
TO THE DEVICE   55 8c 28 00 09
                            ->[..........]<- actual state switch happens within this time range
FROM THE DEVICE                           55 8c 28 00 09
```

So, controlling the device we know only that some state transition **happens after** some point in our code and **happens before** some other point. We are perfectly fine with that because we don't try interacting the device between this points, otherwise we would appear in the Schrodinger cat case. It means that the both following cases **may** happen:

```
(1) TIME ->
TO THE DEVICE   55 8c 28 00 09
                            ->[..........]<- actual state switch happens within this time range
FROM THE DEVICE                           55 8c 28 00 09
```
```
(2) TIME ->
TO THE DEVICE   55 8c 28 00 09
                            ->[..........]<- actual state switch happens within this time range
FROM THE DEVICE                  d0 ...   55 8c 28 00 09
```

and the following case obviously **must not** happen:
```
(3) TIME ->
TO THE DEVICE   55 8c 28 00 09
                            ->[..........]<- actual state switch happens within this time range
FROM THE DEVICE                           55 8c 28 00 09     d0 ...
```

Unfortunately, we see exactly the third case. The following is real transmission with my comments
```
      Host                Device
Please, set SDRAM_HOST_READ_GO to 0
 -> 55 8c 28 00 09
                     Done
                      <- 55 8c 28 00 09
(We are paranoids, lets check what is the SDRAM_HOST_READ_GO value)
What is SDRAM_HOST_READ_GO?
 -> 55 0c 28 00 89
                     It is disabled
                      <- 55 0c 28 00 89
Nice, lets talk about something else
 -> 55 8e 11 00 f4
                      <- 55 8e 11 00 f4
 -> 55 88 00 00 dd
                      <- 55 88 00 00 dd
 -> 55 8e 0c 00 ef
                      <- d0 1b 5a a0 00
       WTF???
```

**Comment by desowin** [comment]
You seem to miss the point completely. There is absolutely no need to set `SDRAM_HOST_READ_GO` to `0` and doing so it actually harmful. The only way you can see if something happens after and before some other point in time is to create stuffed packet, this is what `HF0_FIRST` and `HF0_LAST` are for.

Host writes have higher priority than device writes on FTDI bus. Therefore any register write will interrupt mid-flight any ongoing device write, i.e. the `0xD0` will effectively get interrupted and will continue only after the read is finished. The register write is processed few clock cycles after appearing on FTDI interface. The register write response will be sent immediately after the interrupted `0xD0` is finished - see `BusInterleave` class.

`SDRAM_HOST_READ_GO` set to `0` only prevents new data from entering `sdram_fifo` in `SDRAM_Host_Read`. But it does not magically make the data already sitting in `sdram_fifo` go back into DRAM. It cannot really go back to DRAM, because `SDRAM_Host_Read` is DRAM reader and the actual capture is the DRAM writer. Whole DRAM is essentially a huge ring buffer with one reader and one writer.

It is very unlikely (I not sure if at all possible, but I don't think there is point in checking that - that is, if it is possible the "time window" you need to hit is a single specific 100 MHz clock cycle; if it is not possible, then there is no such "time window" at all; note that FTDI bus is clocked at 60 MHz) that you make the `SDRAM_HOST_READ_GO` write to `0` to actually change the register in the same clock cycle that the `0xD0` ends. Therefore there will virtually always be pending partial data in the `sdram_fifo` after the current `0xD0` packets (interrupted by host write) finishes. That data is whatever entered the fifo before `SDRAM_HOST_READ_GO` changed to `0`. The pending partial data will always leave the system after the `FLUSH_TIMEOUT` because if you set `SDRAM_HOST_READ_GO` to `0` no new data will enter `sdram_fifo`.

`FLUSH_TIMEOUT` is pretty high because sending data in larger bursts gives more throughput. It is so high, that for sure you can do multiple other register accesses in between. `sdram_fifo.level` is not accessible via register, but you can rather easily add it to `SDRAM_Host_Read`, e.g.:
```
self._debug_fifo_level = CSRStatus(8)
self.comb += self._debug_fifo_level.status.eq(sdram_fifo.level)
```
just below `sdram_fifo` is instatiated. You can then build the bitstream and keep reading the status register to see that it indeed does not increase after you set `SDRAM_HOST_READ_GO` to `0`. It will just remain at constant level and then decrease. When it gets to 0 and `SDRAM_HOST_READ_GO` is `0` there won't be any more `0xD0` data.

All state machines operate in parallel. During active capture, `sdram_fifo` can be pretty much always non-empty. This is fine, because the burst is only reading as many 16-bit words as was available in `sdram_fifo` at the time `host_write_fsm` decided to burst out the data (and thus write the `0xD0` header).

You might be applying sequential programming logic to heavily parallel processes that happen in gateware (and thus arrive with wrong expectations towards gateware operation).

TL;DR just don't set `SDRAM_HOST_READ_GO` to `0` because it is pointless. To stop capture simply set `CSTREAM_CFG` to `0` and wait for `HF0_LAST` instead. This will fully flush all the fifos in gateware, and the received packet stream will contain data captured only up to the packet that started on the bus before the `CSTREAM_CFG` was set to `0`.
